### PR TITLE
fix: [CDS-85881]: Improve template variables rendering performance

### PR DESCRIPTION
**Pull request by pankajsingh-harness**
### Summary
**Issue:** Form field value takes time to update when there are large number of variable fields.

**Fix:**  Added some conditions to improve performance

- Return early if the field need not be shown.
- Do not iterate over the formik values to find the index if the template variable name and formik variable name is the same. As we are comparing both with the name only inside the loop itself. So check before iterating. If they are the same do not iterate.
- Do not call the function to get options for the allowed value field if the allowed value field need not be shown

There are also 2 minor fixes in the PR found while fixing the main issue:

1. Do not show the variable headers if there are no template variables to show. Previously initial values were used to show the header and template value was used to show the list of variables. (Changed the condition to show the header)
2. Wrong allowed values were calculated for the template variable. Formik variable index was used to get the value from the template variable list. Update the index to the template index from formik index.

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
- MessageMetadata: `trigger messagecheck`
- GitLeaks: `trigger gitLeaks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

### CODE REVIEW

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet.tsx [hunk 1]
```diff
@@ -133,25 +141,57 @@ function CustomVariableInputSetBasic(props: ConectedCustomVariableInputSetProps)
         </section>
       )}
       {template?.variables?.map?.((variable, templateIndex) => {
-        // find Index from values, not from template variables
-        // because the order of the variables might not be the same
-        const index = formikVariables.findIndex((fVar: AllNGVariables) => variable.name === fVar.name)
-
-        const selectOpt = defaultTo(getAllowedValuesFromTemplate(template, `variables[${index}].value`), [])
         const value = defaultTo(variable.value, '')
+
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet.tsx:145]: Return early before doing any processing if need not to show the field itself.

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet.tsx [hunk 2]
```diff
@@ -133,25 +141,57 @@ function CustomVariableInputSetBasic(props: ConectedCustomVariableInputSetProps)
...
         if (getMultiTypeFromValue(value as string) !== MultiTypeInputType.RUNTIME) {
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet.tsx:146]: Return early before doing any processing if need not to show the field itself.

**Comment by pankajsingh-harness** [comment]
trigger messagecheck

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet.tsx [hunk 3]
```diff
@@ -133,25 +141,57 @@ function CustomVariableInputSetBasic(props: ConectedCustomVariableInputSetProps)
...
           return
         }
-        const variableFromAllValues = get(allValues, 'variables', []).find(
-          (fVar: AllNGVariables) => variable.name === fVar.name
-        )
-        const description = defaultTo(variableFromAllValues?.description, '')
+
+        // find Index from values, not from template variables
+        // because the order of the variables might not be the same
+        let index = templateIndex
```
Comment (vikrant-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet.tsx:152]: if template variables order doesn't match with formik variables index this templateIndex won't work here ?

File: src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/__tests__/KubernetesServiceSpecInputForm.test.tsx [hunk 4]
```diff
@@ -661,7 +662,17 @@ describe('VARIABLES', () => {
...
+    expect(getByText('name')).toBeInTheDocument()
+    expect(getByText('description')).toBeInTheDocument()
+    expect(getByText('valueLabel')).toBeInTheDocument()
+
+    // Check for the variables
+    expect(getByText('testvar1')).toBeInTheDocument()
+    expect(getByText('common.optionalLabel')).toBeInTheDocument()
+    expect(getByText('String')).toBeInTheDocument()
+    expect(container.querySelector('[data-icon="description"]')).toBeInTheDocument()
+    expect(container.querySelector('input[value="somedefaultvalue"]')).toBeInTheDocument()
```
Comment (chetanvekariya-harness) [inline src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/__tests__/KubernetesServiceSpecInputForm.test.tsx:675]: We can use findByDisplayValue/getByDisplayValue here, I guess.

**Review by chetanvekariya-harness — APPROVED** [review]

**Review by Sagar-Rajput7 — APPROVED** [review]

**Pull request merged**